llama/model_cn_comment.py

```diff
@@ -490,39 +505,51 @@ def __init__(self, params: ModelArgs):
 
     @torch.inference_mode()
     def forward(self, tokens: torch.Tensor, start_pos: int):
-        """Perform a forward pass through the Transformer model.
+        """执行 Transformer 模型的前向过程.
 
         Args:
-            tokens (torch.Tensor): Input token indices.
-            start_pos (int): Starting position for attention caching.
+            tokens (torch.Tensor): 输入的 token 索引.
+            start_pos (int): Attention caching 的起始位置.
 
         Returns:
-            torch.Tensor: Output logits after applying the Transformer model.
+            torch.Tensor: Transformer 模型输出的 logits.
 
         """
         _bsz, seqlen = tokens.shape
+        # 根据输入的 token 索引计算对应的 embedding
         h = self.tok_embeddings(tokens)
         self.freqs_cis = self.freqs_cis.to(h.device)
+        # 根据起始位置和序列长度截取对应的复指数频率张量.
         freqs_cis = self.freqs_cis[start_pos : start_pos + seqlen]
 
         mask = None
+        # 若序列长度大于 1, 则需要计算用于 Attention 的 mask.
         if seqlen > 1:
+            # 首先根据序列长度构建一个 [seqlen, seqlen] 的矩阵, 每一个元素的值均为 -inf.
             mask = torch.full(
                 (seqlen, seqlen), float("-inf"), device=tokens.device
             )
 
+            # 保留该矩阵的上三角部分, 不保留主对角线的值, 即主对角线右上部分的元素保持为 -inf,
+            # 其余元素均为0.
+            # 表示每个 token 只能使用当前 token 以及已经处理过的 token 的信息.
             mask = torch.triu(mask, diagonal=1)
 
-            # When performing key-value caching, we compute the attention scores
-            # only for the new sequence. Thus, the matrix of scores is of size
-            # (seqlen, cache_len + seqlen), and the only masked entries are (i, j) for
-            # j > cache_len + i, since row i corresponds to token cache_len + i.
+            # 由于使用了 key-value caching, 我们只需要为新序列计算 attention scores.
```

Comment:
💎💎💎The statement need to refine. And I think add some newlines will have better read experience.

Reply:
How about:

```
# ┌─────────────────────────────────────────────────────────┐
# │ > mask maxtirx visualization                            │
# │                                                         │
# │                                     hstack              │
# │                                       ↓                 │
# │            ↙ [0][0][0][0][0][0][0][0] | [0][x][x][x][x] │
# │           ↙  [0][0][0][0][0][0][0][0] | [0][0][x][x][x] │
# │ seq_len(5) ← [0][0][0][0][0][0][0][0] | [0][0][0][x][x] │
# │           ↖  [0][0][0][0][0][0][0][0] | [0][0][0][0][x] │
# │            ↖ [0][0][0][0][0][0][0][0] | [0][0][0][0][0] │
# │                                    ↑     ↘  seq_len  ↙  │
# │                               cache_len(8)              │
# │                                                         │
# │ x: denote the -inf value (masked value)                 │
# └─────────────────────────────────────────────────────────┘
```